autogpt_platform/backend/backend/data/credit.py

```diff
@@ -178,42 +332,209 @@ async def spend_credits(
+            raise ValueError("Either session_id or user_id must be provided")
+
+        # Retrieve CreditTransaction
+        find_filter: CreditTransactionWhereInput = {
+            "type": CreditTransactionType.TOP_UP,
+            "isActive": False,
+        }
+        if session_id:
+            find_filter["transactionKey"] = session_id
+        if user_id:
+            find_filter["userId"] = user_id
+
+        # Find the most recent inactive top-up transaction
+        credit_transaction = await CreditTransaction.prisma().find_first_or_raise(
+            where=find_filter,
+            order={"createdAt": "desc"},
+        )
+
+        # This can be called multiple times for one id, so ignore if already fulfilled
+        if not credit_transaction:
+            return
+
+        # Retrieve the Checkout Session from the API
+        checkout_session = stripe.checkout.Session.retrieve(
+            credit_transaction.transactionKey
+        )
+
+        # Check the Checkout Session's payment_status property
+        # to determine if fulfillment should be performed
+        if checkout_session.payment_status in ["paid", "no_payment_required"]:
+            await self._enable_transaction(
+                transaction_key=credit_transaction.transactionKey,
+                user_id=credit_transaction.userId,
+                metadata=Json({"checkout_session": checkout_session}),
+            )
+
+    async def get_credits(self, user_id: str) -> int:
+        balance = await self._get_credits(user_id)
+        grant_amount = await self.get_credit_grant(user_id)
+
+        # Apply grants for each missing month
+        for grant_date in await self._get_missing_grant_months(user_id):
+            try:
+                new_balance = balance + grant_amount
+                await CreditTransaction.prisma().create(
+                    data={
+                        "transactionKey": f"MONTHLY-CREDIT-TOP-UP-{grant_date}",
+                        "userId": user_id,
+                        "amount": grant_amount,
+                        "runningBalance": new_balance,
+                        "type": CreditTransactionType.TOP_UP,
+                        "metadata": Json({}),
+                        "isActive": True,
+                        "createdAt": datetime.combine(grant_date, datetime.min.time()),
+                    }
+                )
+                balance = new_balance
+            except UniqueViolationError:
+                continue
+
+        return balance
+
+    async def _get_missing_grant_months(self, user_id: str) -> list[date]:
+        """
+        Get list of missing grant months after the latest grant.
+        If no grant is found, return only the current month.
+        """
+        cur_date = self.time_now().date().replace(day=1)
+
+        # Find latest grant using transaction key pattern
+        latest_grant = await CreditTransaction.prisma().find_first(
```

Comment:
this will cause the full transaction lookup for the user through the history, and this happens on each get_balance command